Fabric.Identity.API/Services/ClaimsService.cs

```diff
@@ -95,37 +94,57 @@ public string GetEffectiveUserId(ClaimsResult claimInformation)
             string userId = null;
             if (this.IsExternalTokenAzureAD(claimInformation.SchemeItem))
             {
-                userId = AzureADSubjectId(claimInformation);
-            }
-
-            if(userId == null)
-            {
-                userId = claimInformation.UserId;
+                userId = AzureADSubjectId(claimInformation.Claims);
             }
 
-            return userId;
+            return userId ?? claimInformation.UserId;
         }
 
-        private string AzureADSubjectId(ClaimsResult claimInformation) => 
-            claimInformation.Claims.FirstOrDefault(x => x.Type == AzureActiveDirectoryJwtClaimTypes.OID || x.Type == AzureActiveDirectoryJwtClaimTypes.OID_Alternative)?
+        private static string AzureADSubjectId(IEnumerable<Claim> claims) => 
+            claims.FirstOrDefault(x => x.Type == AzureActiveDirectoryJwtClaimTypes.OID || x.Type == AzureActiveDirectoryJwtClaimTypes.OID_Alternative)?
                             .Value;
 
-        private ClaimsResult GenerateNewClaimsResult(AuthenticateInfo info, AuthorizationRequest context)
+        private async Task<ClaimsResult> GenerateNewClaimsResult(AuthenticateInfo info, AuthorizationRequest context)
         {
             // provider and scheme look the same, but if you see the values
             //  FabricIdentityConstants.AuthenticationSchemes.Azure = "AzureActiveDirectory"
             // you will notice there are 2 different values, one for the provider and the other for the scheme
             var provider = info.Properties.Items["scheme"];
             var schemeItem = info.Properties.Items.FirstOrDefault(i => i.Key == "scheme").Value;
-            var externalUser = info?.Principal;
-            if (externalUser == null)
-            {
-                throw new Exception("External authentication error");
+            var claimsPrincipal = info?.Principal;
+            if (claimsPrincipal == null)
+            {
+                throw new Exception("External authentication error");
             }
 
-            var claims = externalUser.Claims.ToList();
+            var claims = claimsPrincipal.Claims.ToList();
             var userIdClaim = this.GetUserIdClaim(claims);
-            return new ClaimsResult()
+
+            var subjectId = AzureADSubjectId(claims);
+
+            if (!string.IsNullOrEmpty(subjectId))
+            {
+                var externalUser = await _externalIdentityProviderService.FindUserBySubjectId(subjectId);
+                if (externalUser != null)

```

Comment:
This is the second time I seen this code.  Could we create a common function of it?